addons/udes_stock_picking_batch/models/stock_picking_batch.py

```diff
@@ -200,3 +222,701 @@ def _remove_unready_picks(self):
+
+        return user_id
+
+    def _create_batch(self, user_id, picking_type_id, picking_priorities=None, picking_id=None):
+        """
+        Create a batch for the specified user by including only
+        those pickings with the specified picking_type_id and picking
+        priorities (optional).
+        The batch will be marked as ephemeral.
+        In case no pickings exist, return None.
+        """
+        PickingBatch = self.env["stock.picking.batch"]
+        Picking = self.env["stock.picking"]
+
+        if picking_id:
+            picking = Picking.browse(picking_id)
+        else:
+            picking = Picking.search_for_pickings(picking_type_id, picking_priorities)
+
+        if not picking:
+            return None
+
+        picking_type = picking.mapped("picking_type_id")
+        picking_type.ensure_one()
+        if picking_type.u_reserve_pallet_per_picking:
+            max_reservable_pallets = picking_type.u_max_reservable_pallets
+            if len(picking) > max_reservable_pallets:
+                raise ValidationError(
+                    "Only %d pallets may be reserved at a time." % max_reservable_pallets
+                )
+
+        batch = PickingBatch.sudo().create({"user_id": user_id})
+        picking.write({"batch_id": batch.id})
+        batch.check_same_picking_priority(picking)
+        batch.write({"u_ephemeral": True})
+        batch.mark_as_todo()
+
+        return batch
+
+    def close(self):
+        """Unassign incomplete pickings from batches. In case of a
+        non-ephemeral batch then incomplete pickings are moved into a new
+        batch.
+        """
+        for batch in self:
+            # Unassign batch_id from incomplete stock pickings on ephemeral batches
+            batch.filtered(lambda b: b.u_ephemeral).mapped("picking_ids").filtered(
+                lambda sp: sp.state not in ("done", "cancel")
+            ).write({"batch_id": False, "u_reserved_pallet": False})
+
+            # Assign incomplete pickings to new batch
+            _logger.info("Creating continuation batch from %r.", batch.name)
+            pickings = (
+                batch.filtered(lambda b: not b.u_ephemeral)
+                    .mapped("picking_ids")
+                    .filtered(lambda sp: sp.state not in ("done", "cancel"))
+            )
+            _logger.info("Picking ids continuation %r", pickings)
+
+            batch._copy_continuation_batch(pickings)
+
+    def _copy_continuation_batch(self, pickings):
+        """
+        Copy a batch and add the provided pickings.
+        The new batch will be named BATCH/nnnnn-XX where XX is a sequence number
+        which will be incremented or set to '01'.
+        The batch will not be marked as ephemeral.
+        In case no pickings exist, return None.
+        """
+        self.ensure_one()
+
+        if not pickings:
+            return None
+
+        new_name = get_next_name(self, "picking.batch")
```

Comment:
Have we ported `get_next_name`?

Reply:
Copying from Lorenzo's MR, as these were part of his story. I think have missed it at first